File: project/block_manager/services/nodes/pytorch/avgpool2d.py
"""PyTorch AvgPool2D Node Definition"""

from typing import Dict, List, Optional, Any
from ..base import NodeDefinition, NodeMetadata, ConfigField, TensorShape, Framework


class AvgPool2DNode(NodeDefinition):
    """2D Average Pooling layer"""
    
    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="avgpool2d",
            label="AvgPool2D",
            category="basic",
            color="var(--color-primary)",
            icon="ArrowsInSimple",
            description="Average pooling for 2D inputs",
            framework=Framework.PYTORCH
        )
    
    @property
    def config_schema(self) -> List[ConfigField]:
        return [
            ConfigField(
                name="kernel_size",
                label="Kernel Size",
                type="number",
                default=2,
                min=1,
                description="Size of the pooling window"
            ),
            ConfigField(
                name="stride",
                label="Stride",
                type="number",
                default=2,
                min=1,
                description="Stride of the pooling window"
            ),
            ConfigField(
                name="padding",
                label="Padding",
                type="number",
                default=0,
                min=0,
                description="Zero padding on both sides"
            )
        ]
    
    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        if not input_shape or len(input_shape.dims) != 4:
            return None
        
        batch, channels, height, width = input_shape.dims
        kernel_size = int(config.get("kernel_size", 2))
        stride = int(config.get("stride", 2))
        padding = int(config.get("padding", 0))
        
        # Calculate output dimensions
        out_height = ((height + 2 * padding - kernel_size) // stride) + 1
        out_width = ((width + 2 * padding - kernel_size) // stride) + 1
        
        return TensorShape(
            dims=[batch, channels, out_height, out_width],
            description=f"AvgPool2D({kernel_size}x{kernel_size})"
        )
    
    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        # Allow connections from input/dataloader without shape validation
        if source_node_type in ("input", "dataloader"):
            return None
        
        # Empty and custom nodes are flexible
        if source_node_type in ("empty", "custom"):
            return None
        
        # Validate 4D input (N, C, H, W)
        return self.validate_dimensions(
            source_output_shape,
            4,
            "[batch, channels, height, width]"
        )

File: project/frontend/src/lib/nodes/definitions/pytorch/avgpool2d.ts
/**
 * PyTorch AvgPool2D Layer Node Definition
 */

import { NodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField, BlockType } from '../../../types'

export class AvgPool2DNode extends NodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'avgpool2d',
    label: 'AvgPool2D',
    category: 'basic',
    color: 'var(--color-primary)',
    icon: 'ArrowsInSimple',
    description: 'Average pooling for 2D inputs',
    framework: BackendFramework.PyTorch
  }

  readonly configSchema: ConfigField[] = [
    {
      name: 'kernel_size',
      label: 'Kernel Size',
      type: 'number',
      default: 2,
      min: 1,
      description: 'Size of the pooling window'
    },
    {
      name: 'stride',
      label: 'Stride',
      type: 'number',
      default: 2,
      min: 1,
      description: 'Stride of the pooling window'
    },
    {
      name: 'padding',
      label: 'Padding',
      type: 'number',
      default: 0,
      min: 0,
      description: 'Zero padding on both sides'
    }
  ]

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    if (!inputShape) {
      return undefined
    }

    if (inputShape.dims.length !== 4) {
      return undefined
    }

    const [batch, channels, height, width] = inputShape.dims as number[]
    const kernel = (config.kernel_size ?? 2) as number
    const stride = (config.stride ?? 2) as number
    const padding = (config.padding ?? 0) as number

    const outHeight = Math.floor((height + 2 * padding - kernel) / stride) + 1
    const outWidth = Math.floor((width + 2 * padding - kernel) / stride) + 1

    return {
      dims: [batch, channels, outHeight, outWidth],
      description: `AvgPool2D(${kernel}x${kernel})`
    }
  }

  validateIncomingConnection(
    sourceNodeType: BlockType,
    sourceOutputShape: TensorShape | undefined,
    targetConfig: BlockConfig
  ): string | undefined {
    if (sourceNodeType === 'input' || sourceNodeType === 'dataloader') {
      return undefined
    }

    if (sourceNodeType === 'empty' || sourceNodeType === 'custom') {
      return undefined
    }

    return this.validateDimensions(sourceOutputShape, {
      dims: 4,
      description: '[batch, channels, height, width]'
    })
  }
}
